### PR TITLE
feat(observability): utilise new logger in stats-web

### PR DESCRIPTION
**Pull request by forbesus**
ref #436

### CODE REVIEW

File: apps/stats-web/src/app/transactions/[hash]/page.tsx [hunk 1]
```diff
@@ -33,7 +36,7 @@ export async function generateMetadata({ params: { hash } }: TransactionDetailPa
 
 async function fetchTransactionData(hash: string, network: Network["id"]): Promise<TransactionDetail | null> {
   const apiUrl = serverApiUrlService.getBaseApiUrlFor(network);
-  console.log("DEBUG apiUrl", apiUrl);
+  transactionLogger.info(`DEBUG apiUrl: ${apiUrl}`);
```
Comment (ygrishajev) [inline apps/stats-web/src/app/transactions/[hash]/page.tsx]:
```suggestion
```

File: apps/stats-web/src/app/transactions/[hash]/error.tsx [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
 import PageContainer from "@/components/PageContainer";
 import { Title } from "@/components/Title";
 
+const transactionLogger = new LoggerService({ context: "transaction" });
+
 export default function Error({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
   useEffect(() => {
     // Log the error to an error reporting service
-    console.error(error);
+    transactionLogger.error(error);
```
Comment (ygrishajev) [inline apps/stats-web/src/app/transactions/[hash]/error.tsx]:
```suggestion
    transactionLogger.debug(error);
```

File: apps/stats-web/src/app/error.tsx [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
 import PageContainer from "@/components/PageContainer";
 import { Title } from "@/components/Title";
 
+const errorLogger = new LoggerService({ context: "error" });
+
 export default function Error({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
   useEffect(() => {
     // Log the error to an error reporting service
-    console.error(error);
+    errorLogger.error(error);
```
Comment (ygrishajev) [inline apps/stats-web/src/app/error.tsx]:
```suggestion
    errorLogger.debug(error);
```

File: apps/stats-web/src/app/blocks/[height]/errors.tsx [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
 import { useEffect } from "react";
+import { LoggerService } from "@akashnetwork/logging";
+
+const blockErrorLogger = new LoggerService({ context: "block" });
 
 export default function Error({ error, reset }: { error: Error & { digest?: string }; reset: () => void }) {
   useEffect(() => {
     // Log the error to an error reporting service
-    console.error(error);
+    blockErrorLogger.error(error);
```
Comment (ygrishajev) [inline apps/stats-web/src/app/blocks/[height]/errors.tsx]:
```suggestion
    blockErrorLogger.debug(error);
```

File: apps/stats-web/src/lib/copyClipboard.ts [hunk 5]
```diff
@@ -14,9 +18,9 @@ function fallbackCopyTextToClipboard(text: string) {
   try {
     const successful = document.execCommand("copy");
     const msg = successful ? "successful" : "unsuccessful";
-    console.log("Fallback: Copying text command was " + msg);
+    libLogger.info("Fallback: Copying text command was " + msg);
```
Comment (ygrishajev) [inline apps/stats-web/src/lib/copyClipboard.ts]:
```suggestion
    libLogger.debug("Fallback: Copying text command was " + msg);
```

File: apps/stats-web/src/lib/copyClipboard.ts [hunk 6]
```diff
@@ -14,9 +18,9 @@ function fallbackCopyTextToClipboard(text: string) {
...
   } catch (err) {
-    console.error("Fallback: Oops, unable to copy", err);
+    libLogger.error(`Fallback: Oops, unable to copy: ${err}`);
```
Comment (ygrishajev) [inline apps/stats-web/src/lib/copyClipboard.ts]:
```suggestion
    libLogger.debug(`Fallback: Oops, unable to copy: ${err}`);
```

File: apps/stats-web/src/lib/copyClipboard.ts [hunk 7]
```diff
@@ -28,10 +32,10 @@ export const copyTextToClipboard = (text: string) => {
   }
   navigator.clipboard.writeText(text).then(
     () => {
-      console.log("Async: Copying to clipboard was successful!");
+      libLogger.info("Async: Copying to clipboard was successful!");
```
Comment (ygrishajev) [inline apps/stats-web/src/lib/copyClipboard.ts]:
```suggestion
      libLogger.debug("Async: Copying to clipboard was successful!");
```

File: apps/stats-web/src/lib/copyClipboard.ts [hunk 8]
```diff
@@ -28,10 +32,10 @@ export const copyTextToClipboard = (text: string) => {
...
     },
     err => {
-      console.error("Async: Could not copy text: ", err);
+      libLogger.error(`Async: Could not copy text: ${err}`);
```
Comment (ygrishajev) [inline apps/stats-web/src/lib/copyClipboard.ts]:
```suggestion
      libLogger.debug(`Async: Could not copy text: ${err}`);
```

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 9]
```diff
@@ -2,8 +2,6 @@ import { context, trace } from "@opentelemetry/api";
 import { isHttpError } from "http-errors";
 import pino, { Bindings, Logger as PinoLogger, LoggerOptions } from "pino";
 import { gcpLogOptions } from "pino-cloud-logging";
-import pinoFluentd from "pino-fluentd";
-import pretty from "pino-pretty";
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: this is still needed, this is for readability when developing locally

**Comment by forbesus** [comment]
would you check this pr? @ygrishajev

File: apps/stats-web/src/app/blocks/[height]/errors.tsx [hunk 10]
```diff
@@ -1,11 +1,14 @@
 "use client"; // Error components must be Client Components
 
 import { useEffect } from "react";
+import { LoggerService } from "@akashnetwork/logging";
+
+const blockErrorLogger = new LoggerService({ context: "block" });
```
Comment (ygrishajev) [inline apps/stats-web/src/app/blocks/[height]/errors.tsx]: this is generally used for dev. So we need to be able to look it up faster. prob smth like this would make sense more

```suggestion
const blockErrorLogger = new LoggerService({ context: "apps/stats-web/src/app/blocks/[height]/errors.tsx" });
```

wdut?

File: packages/logging/src/servicies/logger/logger.service.ts (hunk 9 above)
Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: actually, `pinoFluentd ` and `pretty` is not working on client side @ygrishajev

Comment (ygrishajev) [reply packages/logging/src/servicies/logger/logger.service.ts]: well, this was the point of the other task to make it work nicely in both envs.
Basically we need to load pretty conditionally, or in the worst case scenario have 2 loggers.

Comment (ygrishajev) [reply packages/logging/src/servicies/logger/logger.service.ts]: i see from the doc it's possible to use it without import https://github.com/pinojs/pino-pretty?tab=readme-ov-file#programmatic-integration

meaning we can add this option conditionally based on the runtime env

**Comment by forbesus** [comment]
all things are completed @ygrishajev

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 11]
```diff
@@ -28,20 +26,17 @@ export class LoggerService implements Logger {
       }
     };
 
-    if (config.STD_OUT_LOG_FORMAT === "pretty") {
-      destinations.push(pretty({ sync: true }));
-    } else {
-      options = gcpLogOptions(options as any) as LoggerOptions;
+    if (typeof window === 'undefined') {
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: it should still respect `STD_OUT_LOG_FORMAT`

**Review by ygrishajev — COMMENTED** [review]
so how does it work with process.env and config within logger?

**Comment by forbesus** [comment]
how about now? @ygrishajev

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 12]
```diff
@@ -29,19 +27,18 @@ export class LoggerService implements Logger {
     };
 
     if (config.STD_OUT_LOG_FORMAT === "pretty") {
-      destinations.push(pretty({ sync: true }));
+      if (typeof window !== 'undefined') {
+        options.transport = {
+          target: 'pino-pretty',
+          options: { colorize: true }
+        };
+        destinations.push(process.stdout);
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: `destinations` is redundant and not needed anymore, right?

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 13]
```diff
@@ -29,19 +27,18 @@ export class LoggerService implements Logger {
     };
 
     if (config.STD_OUT_LOG_FORMAT === "pretty") {
-      destinations.push(pretty({ sync: true }));
+      if (typeof window !== 'undefined') {
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: lets do it like so

```typescript
if (typeof window !== "undefined" && config.STD_OUT_LOG_FORMAT === "pretty") {
      options.transport = {
        target: "pino-pretty",
        options: { colorize: true }
      };
      destinations.push(process.stdout);
    } else {
      options = gcpLogOptions(options as any) as LoggerOptions;
    }
```

in browser it's fine to log raw objects

**Review by ygrishajev — COMMENTED** [review]
so I have `LOG_LEVEL=debug` but it's not picked up by config in the browser. I think config needs to be provided to the LoggerService constructor.

**Comment by forbesus** [comment]
> so I have `LOG_LEVEL=debug` but it's not picked up by config in the browser. I think config needs to be provided to the LoggerService constructor.

i am not sure what this meaning exactly
could you tell me more detail?

**Comment by ygrishajev** [comment]
@forbesus LOG_LEVEL is an env var

```ts
LOG_LEVEL: z.enum(["fatal", "error", "warn", "info", "debug", "trace"]).optional().default("info"),
```

which controls which logs are done. It is a great way to configure logger on the backend, but it wouldn't work in most frontend implementations.

So LoggerService should probably implement both env vars and config object as param in constructor.

**Comment by forbesus** [comment]
how about now? @ygrishajev

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 14]
```diff
@@ -16,7 +16,7 @@ export class LoggerService implements Logger {
 
   private initPino(bindings?: Bindings): Logger {
     let options: LoggerOptions = {
-      level: config.LOG_LEVEL,
+      level: bindings?.logLevel || config.LOG_LEVEL,
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: bindings is what goes to log object with every log, config is used to parametrize pino and logger. so the interface should be smth like

```ts
constructor(options?: LoggerOptions & { context: string })
```

this way existing client code wouldn't break. In the init though we'd need to put context into bindings

**Comment by forbesus** [comment]
please check again @ygrishajev

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 15]
```diff
@@ -13,35 +10,30 @@ export type Logger = Pick<PinoLogger, "info" | "error" | "warn" | "debug">;
...
-
-    let options: LoggerOptions = {
-      level: config.LOG_LEVEL,
+  private initPino(options?: LoggerOptions, bindings?: Bindings): Logger {
+    let loggerOptions: LoggerOptions = {
+      level: bindings?.level || config.LOG_LEVEL,
       mixin: () => {
         const currentSpan = trace.getSpan(context.active());
-        return { ...currentSpan?.spanContext() };
+        return { ...currentSpan?.spanContext(), context: bindings?.context };
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: no need to add it here

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 16]
```diff
@@ -13,35 +10,30 @@ export type Logger = Pick<PinoLogger, "info" | "error" | "warn" | "debug">;
 export class LoggerService implements Logger {
   protected pino: Logger;
 
-  constructor(bindings?: Bindings) {
-    this.pino = this.initPino(bindings);
+  constructor(options?: LoggerOptions & { context: string }) {
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: we need to make sure this is taken into account https://github.com/pinojs/pino/blob/HEAD/docs/api.md#bindings

**Comment by forbesus** [comment]
how about now? @ygrishajev

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 17]
```diff
@@ -34,7 +34,7 @@ export class LoggerService implements Logger {
       ...inputOptions
     };
 
-    if (typeof window === "undefined" && config.STD_OUT_LOG_FORMAT === "pretty") {
+    if (typeof window !== "undefined" && config.STD_OUT_LOG_FORMAT === "pretty") {
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: I'm pretty sure this should stay as is. Why'd you change this?
```suggestion
    if (typeof window === "undefined" && config.STD_OUT_LOG_FORMAT === "pretty") {
```

File: apps/stats-web/src/config/browser-env.config.ts [hunk 18]
```diff
@@ -6,5 +6,6 @@ export const browserEnvConfig = validateStaticEnvVars({
   NEXT_PUBLIC_NODE_ENV: process.env.NEXT_PUBLIC_NODE_ENV,
   NEXT_PUBLIC_BASE_API_TESTNET_URL: process.env.NEXT_PUBLIC_BASE_API_TESTNET_URL,
   NEXT_PUBLIC_BASE_API_SANDBOX_URL: process.env.NEXT_PUBLIC_BASE_API_SANDBOX_URL,
-  NEXT_PUBLIC_BASE_API_MAINNET_URL: process.env.NEXT_PUBLIC_BASE_API_MAINNET_URL
+  NEXT_PUBLIC_BASE_API_MAINNET_URL: process.env.NEXT_PUBLIC_BASE_API_MAINNET_URL,
+  NEXT_PUBLIC_BASE_API_LOG_LEVEL: process.env.NEXT_PUBLIC_LOG_LEVEL
```
Comment (ygrishajev) [inline apps/stats-web/src/config/browser-env.config.ts]: how is this var utilised?

File: packages/logging/src/servicies/logger/logger.service.ts (hunk 17 above)
Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: sorry, i just confused

File: apps/stats-web/src/config/browser-env.config.ts (hunk 18 above)
Comment (forbesus) [reply apps/stats-web/src/config/browser-env.config.ts]: it's not using now

File: packages/logging/src/servicies/logger/logger.service.ts (hunk 17 above)
Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: image
i am getting this error when i use LoggerService 
would you check this? @ygrishajev

Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: how about using `this.pino.child()` instead of `this.pino.setBindings()` ?
```js
setContext(context: string) {
    this.pino.child({ context });

    return this;
  }
```
like this @ygrishajev

Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: and i think making two logger may good choice but it's worst case  because next.js is running on both of ssr and csr

**Comment by forbesus** [comment]
hello @ygrishajev i just fixed this issue
would you check again?
thanks

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 19]
```diff
@@ -35,7 +35,7 @@ export class LoggerService implements Logger {
       ...inputOptions
     };
 
-    if (typeof window === "undefined" && config.STD_OUT_LOG_FORMAT === "pretty") {
+    if (typeof window === "undefined" && config.STD_OUT_LOG_FORMAT === "pretty" && config.NODE_ENV !== "production") {
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: this class shouldn't depend on NODE_ENV, only on env vars specific to it and direct config passed to the constructor.
```suggestion
    if (typeof window === "undefined" && config.STD_OUT_LOG_FORMAT === "pretty") {
```

File: packages/logging/src/config/env.config.ts [hunk 20]
```diff
@@ -6,6 +6,7 @@ dotenv.config();
 
 const envSchema = z.object({
   LOG_LEVEL: z.enum(["fatal", "error", "warn", "info", "debug", "trace"]).optional().default("info"),
+  NODE_ENV: z.enum(["development", "production", "test"]).optional().default("development"),
```
Comment (ygrishajev) [inline packages/logging/src/config/env.config.ts]: this package shouldn't depend on NODE_ENV, only on env vars specific to it and direct config passed to the constructor.

```suggestion
```

File: apps/stats-web/next.config.js [hunk 21]
```diff
@@ -13,6 +13,14 @@ try {
 
 /** @type {import('next').NextConfig} */
 const nextConfig = {
+  webpack: (config, { isServer }) => {
```
Comment (ygrishajev) [inline apps/stats-web/next.config.js]: could you please explain why is this needed?

Comment (forbesus) [reply apps/stats-web/next.config.js]: @ygrishajev 
now the Next.js bundler is interfering with how thread-stream resolves its worker files. because worker threads used in server-side libraries are not bundled correctly, i excluded it

Comment (ygrishajev) [reply apps/stats-web/next.config.js]: what issue was it causing exactly? throwing some kind of exception? at build time? at run time?

Comment (forbesus) [reply apps/stats-web/next.config.js]: image
i am getting this error at runtime

Comment (forbesus) [reply apps/stats-web/next.config.js]: i did some research on it and found a solution by changing `next.config.js` 
@ygrishajev

Comment (ygrishajev) [reply apps/stats-web/next.config.js]: i don't mind this in next.config. I just was curious what is it

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 22]
```diff
@@ -48,13 +48,13 @@ export class LoggerService implements Logger {
   }
 
   setContext(context: string) {
-    this.pino.setBindings({ context });
+    this.pino.child({ context });
```
Comment (ygrishajev) [inline packages/logging/src/servicies/logger/logger.service.ts]: why so?

Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: using `setBindings` had been reported as bug
https://github.com/pinojs/pino/issues/2021
@ygrishajev

**Comment by ygrishajev** [comment]
@forbesus I'm getting this error on the backend

```
stats-web-1  |  ⨯ Error: Cannot find module 'pino'
stats-web-1  | Require stack:
stats-web-1  | - /app/apps/stats-web/.next/server/app/_not-found/page.js
stats-web-1  |     at Object.pino (/app/apps/stats-web/.next/server/app/_not-found/page.js:176:18)
stats-web-1  |     at __webpack_require__ (/app/apps/stats-web/.next/server/webpack-runtime.js:33:43)
stats-web-1  |     at eval (../../packages/logging/src/servicies/logger/logger.service.ts:7:62)
stats-web-1  |     at (ssr)/../../packages/logging/src/servicies/logger/logger.service.ts (/app/apps/stats-web/.next/server/app/_not-found/page.js:775:1)
stats-web-1  |     at __webpack_require__ (/app/apps/stats-web/.next/server/webpack-runtime.js:33:43)
stats-web-1  |     at eval (../../packages/logging/src/index.ts:5:90)
stats-web-1  |     at (ssr)/../../packages/logging/src/index.ts (/app/apps/stats-web/.next/server/app/_not-found/page.js:764:1)
stats-web-1  |     at __webpack_require__ (/app/apps/stats-web/.next/server/webpack-runtime.js:33:43)
stats-web-1  |     at eval (./src/app/error.tsx:9:79)
stats-web-1  |     at (ssr)/./src/app/error.tsx (/app/apps/stats-web/.next/server/app/_not-found/page.js:324:1)
stats-web-1  |     at Object.__webpack_require__ [as require] (/app/apps/stats-web/.next/server/webpack-runtime.js:33:43)
stats-web-1  |     at JSON.parse (<anonymous>)
stats-web-1  |     at JSON.parse (<anonymous>)
```

File: packages/logging/src/servicies/logger/logger.service.ts [hunk 23]
```diff
@@ -60,8 +60,10 @@ export class LoggerService implements Logger {
     }
 
     try {
-      // eslint-disable-next-line @typescript-eslint/no-var-requires
-      return require("pino-pretty")({ colorize: true, sync: true });
```
Comment (forbesus) [inline packages/logging/src/servicies/logger/logger.service.ts]: in order to prevent loading server-specific libraries on the client side (browser), we should set `if (typeof window === "undefined")` in front of `require`  @ygrishajev

Comment (ygrishajev) [reply packages/logging/src/servicies/logger/logger.service.ts]: yea, take a look few lines higher ;)

Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: return is working on runtime
at compile time, require is working csr and ssr , and return is not working
pino-pretty should be worked with ssr
but `pino-pretty` is always importing

so we should change this code like this
```
private getPrettyIfPresent(): PinoPretty.PrettyStream | undefined {
    if (typeof window !== "undefined" || LoggerService.config.STD_OUT_LOG_FORMAT !== "pretty") {
      return;
    } else {
      try {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        return require("pino-pretty")({ colorize: true, sync: true });
      } catch (e) {
        this.debug({ context: LoggerService.name, message: "Failed to load pino-pretty", error: e });
        /* empty */
      }
    }
  }```

Comment (forbesus) [reply packages/logging/src/servicies/logger/logger.service.ts]: @ygrishajev

**Comment by forbesus** [comment]
moved to #516
